File: fixtures/kotlin/type_inference/discriminated_union/discriminated_union.kt
sealed class Result {
    data class Success(val value: String) : Result()
    data class Failure(val message: String) : Result()
    object Pending : Result()
}

fun describe(result: Result): String {
    return when (result) {
        is Result.Success -> result.value
        is Result.Failure -> result.message
        is Result.Pending -> "pending"
    }
}

fun handleIs(result: Result): String {
    if (result is Result.Success) {
        return result.value
    }
    return "not success"
}

fun main() {
    println(describe(Result.Success("done")))
    println(handleIs(Result.Failure("oops")))
}

File: fixtures/dart/type_inference/discriminated_union/discriminated_union.dart
sealed class Result {
  const Result();
}

class Success extends Result {
  final String value;
  Success(this.value);
}

class Failure extends Result {
  final String message;
  Failure(this.message);
}

String describe(Result result) {
  if (result is Success) {
    return result.value;
  }
  if (result is Failure) {
    return result.message;
  }
  return 'pending';
}

void main() {
  print(describe(Success('done')));
  print(describe(Failure('oops')));
}
